include/graphblas/algorithms/conjugate_gradient.hpp

```diff
@@ -215,218 +249,330 @@ namespace grb {
+				grb::operators::add< IOType >, grb::operators::mul< IOType >,
+				grb::identities::zero, grb::identities::one
+			>,
+			class Minus = operators::subtract< IOType >,
+			class Divide = operators::divide< IOType >,
+			typename RSI, typename NZI, Backend backend
+		>
+		grb::RC conjugate_gradient(
+			grb::Vector< IOType, backend > &x,
+			const grb::Matrix< NonzeroType, backend, RSI, RSI, NZI > &A,
+			const grb::Vector< InputType, backend > &b,
+			const size_t max_iterations,
+			ResidualType tol,
+			size_t &iterations,
+			ResidualType &residual,
+			grb::Vector< IOType, backend > &r,
+			grb::Vector< IOType, backend > &u,
+			grb::Vector< IOType, backend > &temp,
+			const Ring &ring = Ring(),
+			const Minus &minus = Minus(),
+			const Divide &divide = Divide()
+		) {
+			// static checks
+			static_assert( std::is_floating_point< ResidualType >::value,
+				"Can only use the CG algorithm with floating-point residual "
+				"types." ); // unless some different norm were used: issue #89
+			static_assert( !( descr & descriptors::no_casting ) || (
+					std::is_same< IOType, ResidualType >::value &&
+					std::is_same< IOType, NonzeroType >::value &&
+					std::is_same< IOType, InputType >::value
+				), "One or more of the provided containers have differing element types "
+				"while the no-casting descriptor has been supplied"
+			);
+			static_assert( !( descr & descriptors::no_casting ) || (
+					std::is_same< NonzeroType, typename Ring::D1 >::value &&
+					std::is_same< IOType, typename Ring::D2 >::value &&
+					std::is_same< InputType, typename Ring::D3 >::value &&
+					std::is_same< InputType, typename Ring::D4 >::value
+				), "no_casting descriptor was set, but semiring has incompatible domains "
+				"with the given containers."
+			);
+			static_assert( !( descr & descriptors::no_casting ) || (
+					std::is_same< InputType, typename Minus::D1 >::value &&
+					std::is_same< InputType, typename Minus::D2 >::value &&
+					std::is_same< InputType, typename Minus::D3 >::value
+				), "no_casting descriptor was set, but given minus operator has "
+				"incompatible domains with the given containers."
+			);
+			static_assert( !( descr & descriptors::no_casting ) || (
+					std::is_same< ResidualType, typename Divide::D1 >::value &&
+					std::is_same< ResidualType, typename Divide::D2 >::value &&
+					std::is_same< ResidualType, typename Divide::D3 >::value
+				), "no_casting descriptor was set, but given divide operator has "
+				"incompatible domains with the given tolerance type."
+			);
+			static_assert( std::is_floating_point< ResidualType >::value,
+				"Require floating-point residual type."
+			);
+
+			// create a dummy preconditioner and buffer that will never be used
+			std::function<
+				grb::RC( grb::Vector< IOType >&, const grb::Vector< IOType >& )
+			> dummy_preconditioner =
+				[](grb::Vector< IOType >&, const grb::Vector< IOType >&) -> grb::RC {
+					return grb::FAILED;
+				};
```

Comment:
Error code? (If at some point the code flow changes so that it actually does call the preconditioner even though it should not, then the FAILED error code will propagate)